### PR TITLE
fix(treeview):multiple modifications and optimizations

### PR DESCRIPTION
**Pull request by wildwind2000**
1.Fix change parameter from program not correct affect node state.(Value,Open,Active,Openall) 
2.Change parameter from program(Open,Active,Value) should not trigger event. 
3.Optimize search execution efficiency.

### CODE REVIEW

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 1]
```diff
@@ -452,195 +496,130 @@ private bool FilterTreeItem(TItem item, string? search, Func<TItem, string> item
...
             {
-                var itemKeys = CombineItemKeys(Items);
-                if (_oldItemsKeys != itemKeys)
-                {
-                    _oldItemsKeys = itemKeys;
-
-                    BuildTree(Items, default);
-                }
+                ExcludedKeys = GetExcludeKeys();
+                ComputedItems = Items.Where(r => !ExcludedKeys.Contains(ItemKey.Invoke(r))).ToList();
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:547]:
```c#
if (_preSearch != Search)
{
  _prevSearch = Search;

  // logic
}
```

**Review by capdiem — CHANGES_REQUESTED** [review]

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 2]
```diff
@@ -1,14 +1,15 @@
-﻿namespace BlazorComponent
+﻿using System.Reflection;
+using System.Text;
+
+namespace BlazorComponent
```
[the remark this answers is not in the document]
Comment (wildwind2000) [reply src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs]: ok

[the remark this answers is not in the document]
Comment (wildwind2000) [reply src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs]: ok

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs (hunk 1 above)
Comment (wildwind2000) [reply src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs]: ok

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 3]
```diff
@@ -117,40 +115,22 @@ public bool IsDark
                 return CascadingIsDark;
             }
         }
+        public Dictionary<TKey, NodeState<TItem, TKey>> Nodes { get; private set; } = [];
+        HashSet<TKey>? ExcludedKeys;
+        List<TItem>? ComputedItems;
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:120]: Please use underscored lower camel case naming, move to line 10.

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 4]
```diff
@@ -211,7 +191,25 @@ public bool IsSelected(TKey key)
...
+                }
+                else if (!nodeState.IsActive && Active?.Contains(key) == true)
+                {
+                    Active.Remove(key);
+                }
+                _oldActive = [.. Active ?? []];
+            }
+        }
+
+
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:211]:
```suggestion

```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 5]
```diff
@@ -277,48 +285,74 @@ public async Task EmitOpenAsync()
...
+            => UpdateSelected(key, isSelected, false, []);
 
-        private void UpdateSelectedByValue(TKey key, bool isSelected)
-            => UpdateSelected(key, isSelected, true);
+        private void UpdateSelectedByValue(TKey key, bool isSelected, HashSet<TKey> visited)
+            => UpdateSelected(key, isSelected, true, visited);
 
-        private void UpdateSelected(TKey key, bool isSelected, bool updateByValue)
+        /// <summary>
+        ///
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:294]:
```suggestion
        /// Update the selection state of node.
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 6]
```diff
@@ -373,12 +409,14 @@ private void UpdateParentSelected(TKey? parent, bool isIndeterminate = false)
                         nodeState.IsIndeterminate = true;
                     }
                 }
-
-                UpdateParentSelected(nodeState.Parent, nodeState.IsIndeterminate);
+                visited.Add(parent);
+                if (SelectionType != SelectionType.Leaf) // value not contain parent when selection type is Leaf
+                    store.Add(nodeState);
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:414]:
```suggestion

                // value not contain parent when selection type is Leaf
                if (SelectionType != SelectionType.Leaf)
                {
                    store.Add(nodeState);
                }
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 7]
```diff
@@ -452,195 +491,131 @@ private bool FilterTreeItem(TItem item, string? search, Func<TItem, string> item
...
-            return !string.IsNullOrEmpty(Search) && ExcludedItems.Contains(key);
+            return !string.IsNullOrEmpty(Search) && ExcludedKeys.Contains(key);
         }
 
+        private bool IsItemsChanged() {
+            var newKeys = CombineItemKeys(Items);
+            if (_oldItemsKeys == newKeys) return false;
+            _oldItemsKeys = newKeys;
+            return true;
+        }
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:502]:
```suggestion
        }

```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 8]
```diff
@@ -452,195 +491,131 @@ private bool FilterTreeItem(TItem item, string? search, Func<TItem, string> item
...
         private string CombineItemKeys(IList<TItem> list)
         {
-            string keys = string.Empty;
-
-            if (list == null || list.Count == 0)
-            {
-                return keys;
-            }
+            List<string> li = [];
+            ExploreForKeys(list,li,1);
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:506]:
```suggestion
            ExploreForKeys(list, li, 1);
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 9]
```diff
@@ -452,195 +491,131 @@ private bool FilterTreeItem(TItem item, string? search, Func<TItem, string> item
...
+            if (list == null || list.Count == 0) return;
             foreach (var item in list)
             {
-                var key = ItemKey(item);
-                keys += key.ToString();
-
+                store.Add(ItemKey(item).ToString() ?? "");
                 var children = ItemChildren(item);
-                if (children is not null)
+                if (children is not null && children.Count>0)
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:517]:
```suggestion
                if (children is not null && children.Count > 0)
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 10]
```diff
@@ -117,40 +115,22 @@ public bool IsDark
                 return CascadingIsDark;
             }
         }
+        public Dictionary<TKey, NodeState<TItem, TKey>> Nodes { get; private set; } = [];
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:118]:
```suggestion

        public Dictionary<TKey, NodeState<TItem, TKey>> Nodes { get; private set; } = [];
```

**Review by capdiem — CHANGES_REQUESTED** [review]

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 11]
```diff
@@ -491,7 +489,7 @@ private bool FilterTreeItem(TItem item, string? search, Func<TItem, string> item
 
         public bool IsExcluded(TKey key)
         {
-            return !string.IsNullOrEmpty(Search) && ExcludedKeys.Contains(key);
+            return !string.IsNullOrEmpty(Search) && _excludedKeys?.Contains(key) == true;
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:492]:
```suggestion
            return !string.IsNullOrEmpty(Search) && _excludedKeys?.Contains(key) is true;
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 12]
```diff
@@ -389,7 +383,7 @@ private void UpdateParentSelected(TKey? parent, bool isIndeterminate,List<NodeSt
                 else
                 {
                     var children = Nodes
-                                   .Where(r => nodeState.Children.Contains(r.Key))
+                                   .Where(r => nodeState.Children?.Contains(r.Key) == true)
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:386]:
```suggestion
                                   .Where(r => nodeState.Children?.Contains(r.Key) is true)
```

File: src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs [hunk 13]
```diff
@@ -7,6 +7,8 @@ public partial class BTreeview<TItem, TKey> : ITreeview<TItem, TKey> where TKey
         private HashSet<TKey> _oldActive = [];
         private HashSet<TKey> _oldOpen = [];
         private bool _oldOpenAll;
+        HashSet<TKey>? _excludedKeys;
+        List<TItem>? _computedItems;
```
Comment (capdiem) [inline src/Component/BlazorComponent/Components/Treeview/BTreeview.razor.cs:11]:
```suggestion
        private HashSet<TKey>? _excludedKeys;
        private List<TItem>? _computedItems;
```

**Review by capdiem — CHANGES_REQUESTED** [review]

**Review by capdiem — APPROVED** [review]
Thank you so much for your PR! Your work is greatly appreciated!

**Pull request merged**